src/feature/dirauth/process_descs.c

```diff
@@ -221,50 +218,31 @@ dirserv_load_fingerprint_file(void)
         add_status = FP_INVALID;
     }
 
-    /* Check fingerprint type by key length, and if it's RSA or ed25519. */
-    int is_ed25519 = (strlen(fingerprint) == BASE64_DIGEST256_LEN);
-    int is_rsa = (strlen(fingerprint) == HEX_DIGEST_LEN);
-    bool is_valid_fpr = true;
+    /* Check if fingerprint is RSA or ed25519 by verifying it. */
+    int ed25519_not_ok, rsa_not_ok;
 
-    /* We only accept ed25519 or RSA keys */
-    if (!is_ed25519 && !is_rsa) {
-      is_valid_fpr = false;
+    /* Attempt to add the RSA key. */
+    rsa_not_ok = add_rsa_fingerprint_to_dir(fingerprint, fingerprint_list_new,
+                                            add_status);
 
-      log_notice(LD_CONFIG, "Unrecognized fingerprint type (nickname '%s', "
-                 "fingerprint %s). Skipping.", nickname, fingerprint);
-      continue;
-    }
-
-    /* Decode ed25519 keys */
+    /* Check ed25519 key. We check the size to prevent buffer overflows.
+     * If valid, attempt to add it, */
     ed25519_public_key_t ed25519_pubkey_tmp;
-    if (is_ed25519 && digest256_from_base64((char *) ed25519_pubkey_tmp.pubkey,
-                                            fingerprint) < 0) {
-      is_valid_fpr = false;
-    }
-
-    /* Decode RSA keys */
-    char digest_tmp[DIGEST_LEN];
-    if (is_rsa &&
-        base16_decode(digest_tmp, sizeof(digest_tmp), fingerprint,
-                      HEX_DIGEST_LEN) != sizeof(digest_tmp)) {
-      is_valid_fpr = false;
+    if (strlen(fingerprint) == BASE64_DIGEST256_LEN &&
+        digest256_from_base64((char *) ed25519_pubkey_tmp.pubkey,
```

Comment:
This is a logic error: digest256_from_base64() returns 0 on success.

Reply:
Your tests didn't pick this up: please add a commit with a test that fails on this code.
And then write a fix to make the test pass.

Reply:
This code would be easier to read, if the decoding was outside the "if" expression.

Reply:
9520e28510e757b54c52b8b2177b63936d16de58

However, I didn't move the decoding outside the "if" expression.

Reply:
I opened a new ticket for the decoding refactor:
https://trac.torproject.org/projects/tor/ticket/30691

And the missing unit test cases:
https://trac.torproject.org/projects/tor/ticket/30692#ticket